README.md
# wi-teams-button

## Teams Button Workspace Integration

Designed for customers who have an existing CVI Subscription.

Adds a Microsoft Teams button on the UI allowing you to join both CVI and WebRTC Meetings.

You can also optionally define up to 6 external cvi providers for a simplified join experience

This has been refactored from a per-device macro to instead run from a central location, ideally as a docker container, and leverages the Webex cloud xAPIs to manage and subscribe to events for your devices.

![img1.png](img/img1.png)

Periodic check of devices occurs every 30 minutes (on the half/hour intervals) to detect if a new device is un/tagged, otherwise devices are re/processed on integration restart.
**Note:** If you untag a device and restart the integration, the panel will **not** be removed from the device. For automated panel cleanup, untag any intended devices and wait for the periodic check to occur.

## Prerequisites

1. Navigate to Workspace Integrations in [Control Hub](https://admin.webex.com/workspaces/integrations)
2. Select `Add integration` then `Upload integration` and provide included manifest.json file - ensure you document the provided credentials
3. Navigate to the newly created Integration and select `Activate` from the `Actions` menu - ensure you document the encoded activation code
3. Add the required Device Tag (default: `wi-room-release`) to each device to be managed by this integration

## Deployment (Local)

1. Clone / Download repository
2. Run `npm install` to add the require dependencies (ensure Node and NPM are installed)
3. Create an `.env` file and include the required variables outlined below.
- Recommend adding `WI_LOGGING=info`, `CONSOLE_LEVEL=debug` and `LOG_DETAILED=true` during initial testing
4. Start the integration using `npm run start`
5. Review the console logs to confirm no errors encountered

## Deployment (Docker)

1. Build and Deploy Docker Container (or deploy to Cloud) - ensure you include the required variables outlined below.
- This integration does not require local device access as all communications are done over Cloud xAPI.

    ```
    > docker build --tag wi-teams-button .
    > docker create --name wi-teams-button \
      -e _ENVIRONMENTAL_VARIABLE_ = _value_ \
      wi-teams-button
    ```

2. Review the logs from the Integration output to confirm no errors encountered

### Environmental Variables

These variables can be individually defined in Docker, or loaded as an `.env` file in the app directory.

| Name | Required | Type | Default | Description
| ---- | ---- | ---- | ------- | -----------
| **Integration Settings**
| CLIENT_ID | **Yes** | string | ` ` | Client Identifier provided during the Integration creation process
| CLIENT_SECRET | **Yes** | string | ` ` | Client Secret provided during the Integration creation process
| **---**
| CODE | no* | string | ` ` | Encoded Activation Code provided during the Integration activation process
| *-- or --*
| OAUTH_URL | no* | string | ` ` | Decoded oAuth URL from the Activation Code
| REFRESH_TOKEN | no* | string | ` ` | Decoded Refresh Token from the Activation Code
| WEBEXAPIS_BASE_URL | no* | string | ` ` | Decoded Webex APIs Base Url from the Activation Code
| APP_URL | no* | ` ` | string | Decoded App Url from the Activation Code
| **---**
| DEVICE_TAG | no | string | `wi-teams-button` | Device Tag used to determine which devices to process
| **Logging Settings**
| LOG_DETAILED | no | bool | `true` | Enable detailed logging
| CONSOLE_LEVEL | no | bool | `info` | Logging level exposed to console
| APP_NAME | no | string | `wi-teams-button` | App Name used for logging service
| SYSLOG_ENABLED | no | bool | `false` | Enable external syslog server
| SYSLOG_HOST | no | string | `syslog` | Destination host for syslog server
| SYSLOG_PORT | no | num | `514` | Destination port for syslog server
| SYSLOG_PROTOCOL | no | str | `udp4` | Destination protocol for syslog server
| SYSLOG_SOURCE | no | str | `localhost` | Host to indicate that log messages are coming from
| LOKI_ENABLED | no | bool | `false` | Enable external Loki logging server
| LOKI_HOST| no | string | `http://loki:3100` | Destination host for Loki logging server
| **HTTP Proxy**
| GLOBAL_AGENT_HTTP_PROXY | no | string | ` ` | Container HTTP Proxy Server (format `http://<ip or fqdn>:<port>`)
| GLOBAL_AGENT_NO_PROXY | no | string | ` ` | Comma Separated List of excluded proxy domains (Supports wildcards)
| **Teams Button**
| TB_CUSTOMER_TENANT | **yes** | bool | ` ` | Customer CVI tenant (include fill domain, such as example@m.webex.com)
| TB_CUSTOMER_TEXT | no | string | `Internal Meeting` | Text shown for Internal meeting option
| TB_BUTTON_TEXT | no | string | `Microsoft Teams` | Text to display on button
| TB_PANEL_ID | no | string | `60` | Panel identifier
| TB_PANEL_TITLE | no | string | `Select Microsoft Teams Meeting` | Title text for Panel
| TB_EXTERNAL_CVI | no | bool | `true` | Disable to not include external cvi option
| TB_PROMPT_TIMEOUT | no | num | `90` | Seconds before prompts are closed
| TB_DEFAULT_TENANT | no | string | `@m.webex.com` | Default tenant for external cvi requests
| TB_HIDE_BUTTON | no | bool | `true` | Hide default teams button from ui
| TB_REBUILD_UI | no | bool | `false` | Enable to rebuild panel and button on Integration restart
| TB_DEFINED_TENANTS | no | json | ` ` |  6 defined buttons for common external CVI providers (see below)
| LOG_UNKNOWN_RESPONSES | no | bool | `false` | Show unknown extension responses in the log (useful in debugging issues)

***Note:** You must either include the encoded Activation Code, or the four individual decoded parameters.

### Defined Tenants

You can optionally include up to 6 pre-defined external CVI providers on this panel allowing your customers easier ability to dial meetings hosted by these providers. 

You need to structure the ENV Parameter as follows (ensure you use a different identifier per entry) 
```
[
  { "id": "btnTenant1", "text": "CompanyA Meeting", "tenant": "company1@m.webex.com" },
  { "id": "btnTenant2", "text": "CompanyC Meeting", "tenant": "teams@company2.onpexip.com" },
]
```

then **remove** the line breaks and **remove** the comma from the last entry, for example:

`TB_DEFINED_TENANTS=[{ "id": "btnTenant1", "text": "CompanyA Meeting", "tenant": "company1@m.webex.com" },{ "id": "btnTenant2", "text": "CompanyC Meeting", "tenant": "teams@company2.onpexip.com" }]`

If your JSON is still not accepted during Integration startup (i.e. it throws an Invalid error), please use a JSON Linter, such as [JSONLint](https://jsonlint.com/) to validate your entries.

## Support

In case you've found a bug, please [open an issue on GitHub](../../issues).

## Disclaimer

This application is provided as a sample only is NOT guaranteed to be bug free and production quality.


CHANGELOG.md
# Changelog

## 0.0.1
- Reset version to 0.0.1
- Added Changelog and embed version display on load


app.js
//
// Teams Button - Workspace Integration
//
// Copyright (c) 2023 Jeremy Willans
// Licensed under the MIT License
//
// App Entrypoint
//

/* eslint-disable no-param-reassign */
const wi = require('workspace-integrations');
const { bootstrap } = require('global-agent');
const schedule = require('node-schedule');
const { cleanEnv, str } = require('envalid');
const logger = require('./src/logger')('app');
const utils = require('./src/utils');
const { TeamsButton, minVersion } = require('./src/teamsButton');
const { name, version } = require('./package.json');

// Process ENV Parameters
const e = cleanEnv(process.env, {
  // Integration Options
  DEVICE_TAG: str({ default: name }),
  WI_LOGGING: str({ default: 'error' }),
  CLIENT_ID: str(),
  CLIENT_SECRET: str(),
  // Integration Credentials
  CODE: str({ default: undefined }),
  OAUTH_URL: str({ default: undefined }),
  REFRESH_TOKEN: str({ default: undefined }),
  WEBEXAPIS_BASE_URL: str({ default: undefined }),
  APP_URL: str({ default: undefined }),
  // Global Agent Proxy
  GLOBAL_AGENT_HTTP_PROXY: str({ default: undefined }),
  GLOBAL_AGENT_NO_PROXY: str({ default: undefined }),
});

// Initialize Proxy Server, if defined.
if (e.GLOBAL_AGENT_HTTP_PROXY) {
  logger.info('invoke global agent proxy');
  bootstrap();
}

// Define WI Configuration from ENV Parameters
const wiConfig = {
  clientId: e.CLIENT_ID,
  clientSecret: e.CLIENT_SECRET,
  activationCode: {
    oauthUrl: e.OAUTH_URL,
    refreshToken: e.REFRESH_TOKEN,
    webexapisBaseUrl: e.WEBEXAPIS_BASE_URL,
    appUrl: e.APP_URL,
  },
  notifications: 'longpolling',
  logLevel: e.WI_LOGGING,
};

// Check and process new device
async function processDevice(i, d, deviceId, deviceObj) {
  let device = deviceObj;
  // Get Device object to obtain status and tag info
  if (!device) {
    try {
      device = await i.devices.getDevice(deviceId);
    } catch (error) {
      logger.warn(`Unable to get device: ${utils.shortName(deviceId)}`);
      logger.debug(deviceId);
      logger.debug(error.message);
      return;
    }
  }
  // Check device has correct tag
  if (!device.tags.includes(e.DEVICE_TAG)) return;
  // Ensure device is online before processing
  if (!device.connectionStatus.match(/^connected/)) return;
  // Ensure device meets version requirement
  if (!utils.versionCheck(minVersion, device.software)) return;
  // Declare Class
  const id = utils.uniqueId(d, deviceId.replace('=', ''));
  d[deviceId] = new TeamsButton(i, id, deviceId);
  logger.info(`${d[deviceId].id}: ${utils.shortName(deviceId)}`);
  logger.info(`${d[deviceId].id}: Creating Instance for ${device.displayName}.`);
  try {
    // configure codec
    await d[deviceId].configureCodec();
  } catch (error) {
    logger.warn(`${d[deviceId].id}: Unable to process Device!`);
    logger.debug(`${d[deviceId].id}: ${error.message}`);
  }
}

// Process devices based on tag
async function processDevices(i, d) {
  try {
    // Get devices from xapi
    const devices = await i.devices.getDevices({ tag: e.DEVICE_TAG });
    if (!devices.length) {
      logger.error('No Matching Devices found!');
      return;
    }

    // Process tagged devices
    await Promise.all(
      devices.map(async (device) => {
        // skip if instance exists
        if (d[device.id]) return;
        await processDevice(i, d, device.id, device);
      }),
    );

    // Remove untagged devices
    const toRemove = Object.keys(d).filter((j) => !devices.map((k) => k.id).includes(j));
    toRemove.forEach(async (item) => {
      logger.info(`${d[item].id}: Device no longer tagged, removing Instance.`);
      try {
        await d[item].removePanel();
      } catch (error) {
        logger.error(`${d[item].id}: Unable to remove panel`);
        logger.debug(`${d[item].id}: ${error.message}`);
      }
      d[item] = null;
      delete d[item];
    });
  } catch (error) {
    logger.warn('Unable to process devices');
    logger.debug(error.message);
  }
}

// Init integration
async function init(json) {
  logger.info(`Teams Button Workspace Integration, v${version}`);
  let i;
  const d = {}; // Device Entities Object
  // Process integration credentials
  if (!e.OAUTH_URL) {
    try {
      wiConfig.activationCode = utils.parseJwt(e.CODE);
    } catch (error) {
      logger.error('Unable to decode token');
      logger.debug(error.message);
      process.exit(1);
    }
  }
  try {
    i = await wi.connect(json);
    i.onError(logger.error);
    i.onAction((action) => logger.info(`Integration action: ${JSON.stringify(action)}`));
    logger.info('Integration activation successful!');
  } catch (error) {
    logger.error('Not able to connect to Integration');
    logger.debug(error.message);
    process.exit(1);
  }

  try {
    // Process devices on startup
    logger.info('--- Processing Devices');
    await processDevices(i, d);

    // Periodically re-process devices to capture tag changes (every 30 mins)
    schedule.scheduleJob('*/30 * * * *', async () => {
      logger.info('--- Periodic Device Processing');
      await processDevices(i, d);
    });

    logger.info('--- Processing WI Subscriptions');
    // Process Device Ready
    i.xapi.status.on('SystemUnit.State.System', async (deviceId, _path, result) => {
      const rr = d[deviceId];
      if (!rr && result === 'Initialized') {
        await processDevice(i, d, deviceId);
      }
    });
    // Process Reboot Event
    i.xapi.event.on('BootEvent', (deviceId, _path, event) => {
      const rr = d[deviceId];
      if (rr) {
        logger.info(`${rr.id}: Device ${event.Action}, Removing Instance.`);
        d[deviceId] = null;
        delete d[deviceId];
      }
    });

    logger.info('--- Processing Teams Button Subscriptions');
    // Process widget action
    i.xapi.event.on('UserInterface.Extensions.Widget.Action', (deviceId, _path, event) => {
      const rr = d[deviceId];
      if (!rr) return;
      rr.handleWidgetAction(event);
    });
    // Process TextInput response
    i.xapi.event.on('UserInterface.Message.TextInput.Response', (deviceId, _path, event) => {
      const rr = d[deviceId];
      if (!rr) return;
      rr.handleTextInputResponse(event);
    });
    // Process Call disconnect event
    i.xapi.event.on('CallDisconnect', (deviceId) => {
      const rr = d[deviceId];
      if (!rr) return;
      rr.handleCallDisconnect();
    });
    // Process TextInput timeout
    i.xapi.event.on('UserInterface.Message.TextInput.Clear', (deviceId, _path, event) => {
      const rr = d[deviceId];
      if (!rr) return;
      rr.handleTextInputTimeout(event);
    });
  } catch (error) {
    logger.warn('Error during device and subscription processing');
    logger.debug(error.message);
  }
}

init(wiConfig);


src/utils.js
//
// Util Module
//

// Parse and return JWT decoded integration credentials
function parseJwt(token) {
  return JSON.parse(Buffer.from(token.split('.')[1], 'base64').toString());
}
exports.parseJwt = parseJwt;

// Abbreviate Device Id
function shortName(deviceId) {
  return `${deviceId.slice(0, 8)}...${deviceId.slice(-8)}`;
}
exports.shortName = shortName;

// Generate Unique Identifier for each Device (used in Logs)
function uniqueId(d, deviceId) {
  const result = deviceId.slice(-4);
  const existing = Object.keys(d).map((j) => d[j].id).includes(result);
  if (!existing) return result;
  return uniqueId(d, deviceId.slice(0, deviceId.length - 1));
}
exports.uniqueId = uniqueId;

function versionCheck(target, source) {
  const reg = /^\D*(?<MAJOR>\d*)\.(?<MINOR>\d*)\.(?<EXTRA>\d*)\.(?<BUILD>\d*).*$/i;
  const x = (reg.exec(source)).groups;
  const y = (reg.exec(target)).groups;
  if (Number(x.MAJOR) > Number(y.MAJOR)) return true;
  if (Number(x.MAJOR) < Number(y.MAJOR)) return false;
  if (Number(x.MINOR) > Number(y.MINOR)) return true;
  if (Number(x.MINOR) < Number(y.MINOR)) return false;
  if (Number(x.EXTRA) > Number(y.EXTRA)) return true;
  if (Number(x.EXTRA) < Number(y.EXTRA)) return false;
  if (Number(x.BUILD) > Number(y.BUILD)) return true;
  if (Number(x.BUILD) < Number(y.BUILD)) return false;
  return false;
}
exports.versionCheck = versionCheck;


src/logger.js
//
// Logger Module
//

/* eslint-disable no-param-reassign */
/* eslint-disable no-underscore-dangle */
const winston = require('winston');
const LokiTransport = require('winston-loki');
const { Syslog } = require('winston-syslog');
const ms = require('ms');
// eslint-disable-next-line object-curly-newline
const { cleanEnv, str, bool, num } = require('envalid');
require('dotenv').config();

let logger = {};

// Process ENV Parameters
const env = cleanEnv(process.env, {
  // Logging Options
  APP_NAME: str({ default: 'wi-teams-button' }),
  SYSLOG_ENABLED: bool({ default: false }),
  SYSLOG_HOST: str({ default: 'syslog' }),
  SYSLOG_PORT: num({ default: 514 }),
  SYSLOG_PROTOCOL: str({ default: 'udp4' }),
  SYSLOG_SOURCE: str({ default: 'localhost' }),
  LOKI_ENABLED: bool({ default: false }),
  LOKI_HOST: str({ default: 'http://loki:3100' }),
  CONSOLE_LEVEL: str({ default: 'info' }),
});

const appName = env.APP_NAME;

const LOG_TIME_DIFF = Symbol('LOG_TIME_DIFF');
// adds data to log event info object
const addTimeDiff = winston.format((info) => {
  const now = Date.now();
  if (!this._lastTimestamp) {
    this._lastTimestamp = now;
    info[LOG_TIME_DIFF] = 0;
  } else {
    const diff = now - this._lastTimestamp;
    this._lastTimestamp = now;
    info[LOG_TIME_DIFF] = diff;
  }

  return info;
});

// render it similar to `debug` library
const msgWithTimeDiff = winston.format((info) => {
  info.message = `${info.message} +${ms(info[LOG_TIME_DIFF])}`;
  return info;
});

function WinstonLogger(component) {
  const labels = {
    app: appName,
  };

  let destinations = 'Console';
  const transports = [
    // printing the logs to console
    new winston.transports.Console({
      format: winston.format.combine(
        addTimeDiff(),
        msgWithTimeDiff(),
        winston.format.errors({ stack: true }),
        winston.format.colorize({
          all: true,
        }),
        winston.format.label({
          label: `[${appName}:${component}]`,
        }),
        winston.format.printf((res) => {
          const time = new Date(Date.now());
          const year = time.getUTCFullYear();
          const month = time.getUTCMonth() + 1;
          const date = time.getUTCDate();
          const hour = time.getUTCHours();
          const min = time.getUTCMinutes();
          const sec = time.getUTCSeconds();

          const timeString = `${year}-${(`0${month}`).slice(-2)}-${(`0${date}`).slice(-2)} ${(`0${hour}`).slice(-2)}:${(`0${min}`).slice(-2)}:${(`0${sec}`).slice(-2)}Z`;
          return `${timeString} ${res.level} ${res.label} ${res.message}`;
        }),
      ),
      level: env.CONSOLE_LEVEL,
    }),
  ];

  if (env.SYSLOG_ENABLED) {
    destinations += ', Syslog';
    transports.push(
      // sending the logs to external syslog server
      new Syslog({
        format: winston.format.combine(
          winston.format.errors({ stack: true }),
          winston.format.label({
            label: `[${appName}:${component}]`,
          }),
          winston.format.printf((res) => `${res.level} ${res.label} ${res.message}`),
        ),
        host: env.SYSLOG_HOST,
        port: env.SYSLOG_PORT,
        protocol: env.SYSLOG_PROTOCOL,
        localhost: env.SYSLOG_SOURCE,
        app_name: appName,
        level: 'debug',
      }),
    );
  }

  if (env.LOKI_ENABLED) {
    destinations += ', Loki';
    transports.push(
      // sending the logs to Loki which will be visualized by Grafana
      new LokiTransport({
        format: winston.format.combine(
          winston.format.errors({ stack: true }),
          winston.format.label({
            label: `[${appName}:${component}]`,
          }),
          winston.format.printf((res) => `${res.level} ${res.label} ${res.message}`),
        ),
        host: env.LOKI_HOST,
        labels,
        level: 'debug',
      }),
    );
  }

  logger = winston.createLogger({
    transports,
  });

  if (component === 'app') logger.info(`Logging destinations enabled: ${destinations}`);

  // Streaming allows it to stream the logs back from the defined transports
  logger.stream = {
    write(message) {
      logger.info(`request: ${message}`);
    },
  };
  return logger;
}

module.exports = WinstonLogger;
